src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -11,28 +11,530 @@
+{
+	return GetForeignConstraintRelationsHelper(relationId, true);
+}
+
+
+/*
+ * GetForeignConstraintRelationsHelper returns the list of oids referenced or
+ * referencing given relation id. It is a helper function for providing results
+ * to public functions ReferencedRelationIdList and ReferencingRelationIdList.
+ */
+static List *
+GetForeignConstraintRelationsHelper(Oid relationId, bool isReferencing)
+{
+	List *foreignKeyList = NIL;
+	bool isFound = false;
+	FRelNode *relationNode = NULL;
+
+	CreateForeignKeyRelationGraph();
+
+	relationNode = (FRelNode *) hash_search(frelGraph->nodeMap, &relationId,
+											HASH_FIND, &isFound);
+
+	if (!isFound)
+	{
+		/*
+		 * If there is no node with the given relation id, that means given table
+		 * is not referencing and does not referenced by any table
+		 */
+		return NIL;
+	}
+	else
+	{
+		List *foreignNodeList = NIL;
+		ListCell *nodeCell = NULL;
+
+		GetConnectedListHelper(relationNode, &foreignNodeList, isReferencing);
+
+		/*
+		 * We need only their OIDs, we get back node list to make their visited
+		 * variable to false for using them iteratively.
+		 */
+		foreach(nodeCell, foreignNodeList)
+		{
+			FRelNode *currentNode = (FRelNode *) lfirst(nodeCell);
+
+			foreignKeyList = lappend_oid(foreignKeyList, currentNode->relationId);
+			currentNode->visited = false;
+		}
+
+		/* set to false separately, since we don't add itself to foreign node list */
+		relationNode->visited = false;
+	}
+
+	return foreignKeyList;
+}
+
+
+/*
+ * GetConnectedListHelper is the function for getting nodes connected (or connecting) to
+ * the given relation. adjacentNodeList holds the result for recursive calls and
+ * by changing isReferencing caller function can select connected or connecting
+ * adjacency list.
+ *
+ */
+static void
+GetConnectedListHelper(FRelNode *node, List **adjacentNodeList, bool isReferencing)
+{
+	ListCell *nodeCell = NULL;
+	List *adjacencyList = NIL;
+
+	node->visited = true;
+
+	if (isReferencing)
+	{
+		adjacencyList = node->backAdjacencyList;
+	}
+	else
+	{
+		adjacencyList = node->adjacencyList;
+	}
+
+	foreach(nodeCell, adjacencyList)
+	{
+		FRelNode *curNode = (FRelNode *) lfirst(nodeCell);
+		if (curNode->visited == false)
+		{
+			*adjacentNodeList = lappend(*adjacentNodeList, curNode);
+			GetConnectedListHelper(curNode, adjacentNodeList, isReferencing);
+		}
+	}
+}
+
+
+/*
+ * CreateForeignKeyRelationGraph creates the foreign key relation graph using
+ * foreign constraint provided by pg_constraint metadata table.
+ */
+static void
+CreateForeignKeyRelationGraph()
+{
+	MemoryContext oldContext;
+	HASHCTL info;
+	uint32 hashFlags = 0;
+	MemoryContext foreignRelationMemoryContext = NULL;
+
+	/* if we have already created the graph, use it */
+	if (IsForeignKeyGraphValid())
+	{
+		return;
+	}
+
+	ClearForeignKeyRelationGraphContext();
+
+	foreignRelationMemoryContext = AllocSetContextCreateExtended(
+		CacheMemoryContext,
+		"FRel Graph Context",
```

Comment:
We generally do not prefer acronyms. Can you go over the code and rename acronyms, especially`FRel` / `frelGraph ` / `FRelNode ` would too hard to understand for a new reader. Also, using `Relation` could be confusing for the readers since `Relation` means a lot in Postgres 😄  So, maybe use `Relationship`? 

maybe use `ForeignConstraintRelationshipGraph` and `ForeignConstraintRelationshipNode`?